cmd/foreach.go

```diff
@@ -0,0 +1,162 @@
+
+The command uses the same filter flags as the 'filter' command to select packages, 
+then executes the specified subcommand for each matched package.`
+
+// getAllowedSubCommands returns the list of allowed subcommands for the foreach command.
+func getAllowedSubCommands() []string {
+	return []string{
+		"build",
+		"check",
+		"clean",
+		"format",
+		"install",
+		"lint",
+		"test",
+		"uninstall",
+	}
+}
+
+func setupForeachCommand() *cobraext.Command {
+	cmd := &cobra.Command{
+		Use:   "foreach [flags] -- <SUBCOMMAND>",
+		Short: "Execute a command for filtered packages",
+		Long:  foreachLongDescription,
+		Example: `  # Run system tests for packages with specific inputs
+  elastic-package foreach --input tcp,udp --parallel 10 -- test system -g`,
+		RunE: foreachCommandAction,
+		Args: cobra.MinimumNArgs(1),
+	}
+
+	// Add filter flags
+	filter.SetFilterFlags(cmd)
+
+	// Add pool size flag
+	cmd.Flags().IntP(cobraext.ForeachPoolSizeFlagName, "", 1, cobraext.ForeachPoolSizeFlagDescription)
+
+	return cobraext.NewCommand(cmd, cobraext.ContextPackage)
+}
+
+func foreachCommandAction(cmd *cobra.Command, args []string) error {
+	poolSize, err := cmd.Flags().GetInt(cobraext.ForeachPoolSizeFlagName)
+	if err != nil {
+		return fmt.Errorf("getting pool size failed: %w", err)
+	}
+
+	if err := validateSubCommand(args[0]); err != nil {
+		return fmt.Errorf("validating sub command failed: %w", err)
+	}
+
+	// Find integration root
+	root, err := packages.MustFindIntegrationRoot()
+	if err != nil {
+		return fmt.Errorf("can't find integration root: %w", err)
+	}
+
+	// reuse filterPackage from cmd/filter.go
+	filtered, err := filterPackage(cmd)
+	if err != nil {
+		return fmt.Errorf("filtering packages failed: %w", err)
+	}
+
+	wg := sync.WaitGroup{}
+	mu := sync.Mutex{}
+	errs := multierror.Error{}
+	successes := 0
+
+	packageChan := make(chan string, poolSize)
+
+	for range poolSize {
+		wg.Add(1)
+		go func(packageChan <-chan string) {
+			defer wg.Done()
+			for packageName := range packageChan {
+				path := filepath.Join(root, "packages", packageName)
+				err := executeCommand(args, path)
+
+				mu.Lock()
+				if err != nil {
+					errs = append(errs, fmt.Errorf("executing command for package %s failed: %w", packageName, err))
+				} else {
+					successes++
+				}
+				mu.Unlock()
+			}
+		}(packageChan)
+	}
+
+	for _, pkg := range filtered {
+		packageChan <- pkg.DirName
+	}
+	close(packageChan)
+
+	wg.Wait()
+
+	logger.Infof("Successfully executed command for %d packages\n", successes)
+	logger.Infof("Errors occurred while executing command for %d packages\n", len(errs))
+
+	if errs.Error() != "" {
+		return fmt.Errorf("errors occurred while executing command for packages: \n%s", errs.Error())
+	}
+
+	return nil
+}
+
+func executeCommand(args []string, path string) error {
+	// Look up the elastic-package binary in PATH
+	execPath, err := exec.LookPath("elastic-package")
```

Comment:
as this is a command run on top of other commands of this binary, could we target directly the commands by its functions, instead of looking for the path binary?
what if the binary is not updated or needs to run specific version? just having some thoughts here on how would be a better aproach without depending of having the binary installed as its own dependency.

Reply:
I think it should be possible to do something like this:
```go
cmd := cmd.RootCmd()
cmd.SetArgs(args)
return cmd.Execute()
```

We could also create our own command that is like `cmd.RootCmd()`, but where we only add the commands with `AddCommand` that we allow.